scripts/test/Muon/grouping_tab/grouping_table_presenter_test.py

```diff
@@ -0,0 +1,295 @@
+import sys
+import six
+from PyQt4 import QtGui
+from Muon.GUI.Common.grouping_table_widget.grouping_table_widget_model import GroupingTableModel
+from Muon.GUI.Common.grouping_table_widget.grouping_table_widget_view import GroupingTableView
+from Muon.GUI.Common.grouping_table_widget.grouping_table_widget_presenter import GroupingTablePresenter
+from Muon.GUI.Common.muon_group import MuonGroup
+from Muon.GUI.Common.muon_data_context import MuonDataContext
+from Muon.GUI.Common import mock_widget
+
+if sys.version_info.major > 2:
+    from unittest import mock
+else:
+    import mock
+
+
+class GroupingTablePresenterTest(unittest.TestCase):
+
+    def setUp(self):
+        self._qapp = mock_widget.mockQapp()
+        # Store an empty widget to parent all the views, and ensure they are deleted correctly
+        self.obj = QtGui.QWidget()
+
+        self.data = MuonDataContext()
+        self.model = GroupingTableModel(data=self.data)
+        self.view = GroupingTableView(parent=self.obj)
+        self.presenter = GroupingTablePresenter(self.view, self.model)
+
+        self.view.warning_popup = mock.Mock()
+
+    def tearDown(self):
+        self.obj = None
+
+    def assert_model_empty(self):
+        self.assertEqual(len(self.model.group_names), 0)
+        self.assertEqual(len(self.model.groups), 0)
+
+    def assert_view_empty(self):
+        self.assertEqual(self.view.num_rows(), 0)
+
+    def add_three_groups_to_table(self):
+        group1 = MuonGroup(group_name="my_group_0", detector_ids=[1])
+        group2 = MuonGroup(group_name="my_group_1", detector_ids=[2])
+        group3 = MuonGroup(group_name="my_group_2", detector_ids=[3])
+        self.presenter.add_group(group1)
+        self.presenter.add_group(group2)
+        self.presenter.add_group(group3)
+
+    # ------------------------------------------------------------------------------------------------------------------
+    # TESTS : Initialization
+    # ------------------------------------------------------------------------------------------------------------------
+
+    def test_that_table_has_three_columns_when_initialized(self):
+        self.assertEqual(self.view.num_cols(), 3)
+
+    def test_that_model_is_initialized_as_empty(self):
+        self.assert_model_empty()
+
+    def test_that_view_is_initialized_as_empty(self):
+        self.assert_view_empty()
+
+    # ------------------------------------------------------------------------------------------------------------------
+    # TESTS : Adding and removing groups
+    # ------------------------------------------------------------------------------------------------------------------
+
+    def test_that_add_group_button_adds_group(self):
+        self.presenter.handle_add_group_button_clicked()
+        self.assertEqual(self.view.num_rows(), 1)
+        self.assertEqual(len(self.model.groups), 1)
+
+    def test_that_remove_group_button_removes_group(self):
+        self.add_three_groups_to_table()
```

Comment:
check we have 3 too

Reply:
I have added a separate test to check that the helper function `add_three_groups_to_table` works correctly as it is used in multiple places.